rs/index/src/ivf/index.rs

```diff
@@ -0,0 +1,111 @@
+use std::cmp::Reverse;
+use std::collections::{BinaryHeap, HashMap};
+
+use utils::l2::L2DistanceCalculator;
+use utils::DistanceCalculator;
+
+use crate::index::Index;
+use crate::utils::{IdWithScore, SearchContext};
+
+pub struct Ivf {
+    // The whole dataset. TODO(tyb0807): reduce the memory footprint.
+    dataset: Vec<Vec<f32>>,
+    // Number of clusters.
+    num_clusters: usize,
+    // Each cluster is represented by a centroid vector. This is all the centroids in our IVF.
+    centroids: Vec<Vec<f32>>,
+    // Inverted index mapping each cluster to the vectors it contains.
+    //   key: centroid index in `centroids`
+    //   value: vector index in `dataset`
+    inverted_lists: HashMap<usize, Vec<usize>>,
+    // Number of probed centroids.
+    num_probes: usize,
+}
+
+impl Ivf {
+    pub fn new(
+        dataset: Vec<Vec<f32>>,
+        centroids: Vec<Vec<f32>>,
+        num_clusters: usize,
+        num_probes: usize,
+    ) -> Self {
+        let mut inverted_lists = HashMap::new();
+        // Assign vectors to nearest centroids
+        for (i, vector) in dataset.iter().enumerate() {
+            let nearest_centroid = Self::find_nearest_centroids(&vector, &centroids, 1);
+            inverted_lists
+                .entry(nearest_centroid[0])
+                .or_insert_with(Vec::new)
+                .push(i);
+        }
+        Self {
+            dataset,
+            num_clusters,
+            centroids,
+            inverted_lists,
+            num_probes,
+        }
+    }
+
+    fn find_nearest_centroids(
+        vector: &Vec<f32>,
+        centroids: &Vec<Vec<f32>>,
+        num_probes: usize,
+    ) -> Vec<usize> {
+        let mut calculator = L2DistanceCalculator::new();
+        let mut distances: Vec<(usize, f32)> = centroids
+            .iter()
+            .enumerate()
+            .map(|(idx, centroid)| {
+                let dist = calculator.calculate(&vector, &centroid);
+                (idx, dist)
+            })
+            .collect();
+        distances.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap());
+        distances.truncate(num_probes);
+        distances.into_iter().map(|(idx, _)| idx).collect()
```

Comment:
do
```
fn find_nearest_centroid(
   &self,
   ...
)
```
Don't need to pass `vector` and `centroids`, since you're keeping it as a member of this instance